test/ast_printing_transform.ts

```diff
@@ -0,0 +1,22 @@
+import * as ts from 'typescript';
+
+/**
+ * Creates a transformer that captures the state of the AST as pretty printed
+ * TypeScript in the provided 'files' map.
+ */
+export function createAstPrintingTransform(files: Map<string, string>) {
+  return (context: ts.TransformationContext): ts.Transformer<ts.SourceFile> => {
+    const visitor: ts.Visitor = (node: ts.Node): ts.Node => {
+      switch (node.kind) {
+        case ts.SyntaxKind.SourceFile:
+          const sf = node as ts.SourceFile;
+          files.set(sf.fileName, ts.createPrinter().printFile(sf));
+          return sf;
+        default:
+          return ts.visitEachChild(node, visitor, context);
```

Comment:
When will this case be hit? As you are passing in a `ts.SourceFile` to `ts.visitNode`, which will then call the `visitor` with a `ts.SourceFile` again...

Reply:
I don't think this will ever be hit, but we need a default case to make the type checker happy.

Reply:
how about throwing?

Reply:
Or even better, don't use `ts.visitNode` at all, call the visitor directly, and type it to receive and return a `ts.SourceFile`. I.e. just return the visitor:
```
return (context: ts.TransformationContext): ts.Transformer<ts.SourceFile> => {
  return (node: ts.SourceFile): ts.Sourcefile => {
    const sf = node as ts.SourceFile;
    files.set(sf.fileName, ts.createPrinter().printFile(sf));
    return sf;  
  };
};
```

Reply:
That works, thanks!